crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+                    key_value: old_leaf.key_value(),
+                },
+                hash: old_leaf.hash,
+                index: 1,
+            },
+        };
+        self.blob.extend(left_leaf_block.to_bytes());
+        self.kv_to_index
+            .insert(left_leaf_block.node.key_value(), left_leaf_block.node.index);
+
+        let right_leaf_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Leaf,
+                dirty: false,
+            },
+            node: Node {
+                parent: Some(0),
+                specific: NodeSpecific::Leaf { key_value },
+                hash,
+                index: 2,
+            },
+        };
+        self.blob.extend(right_leaf_block.to_bytes());
+        self.kv_to_index.insert(
+            right_leaf_block.node.key_value(),
+            right_leaf_block.node.index,
+        );
+
+        self.free_indexes.clear();
+        self.last_allocated_index = 3;
+    }
+
+    fn insert_third_or_later(
+        &mut self,
+        key_value: KvId,
+        hash: Hash,
+        old_leaf: &Node,
+        internal_node_hash: Hash,
+    ) -> Result<(), String> {
+        let new_leaf_index = self.get_new_index();
+        let new_internal_node_index = self.get_new_index();
+
+        let new_leaf_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Leaf,
+                dirty: false,
+            },
+            node: Node {
+                parent: Some(new_internal_node_index),
+                specific: NodeSpecific::Leaf { key_value },
+                hash,
+                index: new_leaf_index,
+            },
+        };
+        self.insert_entry_to_blob(new_leaf_index, new_leaf_block.to_bytes())?;
+
+        let new_internal_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Internal,
+                dirty: false,
+            },
+            node: Node {
+                parent: old_leaf.parent,
+                specific: NodeSpecific::Internal {
+                    left: old_leaf.index,
+                    right: new_leaf_index,
+                },
+                hash: internal_node_hash,
+                index: new_internal_node_index,
+            },
+        };
+        self.insert_entry_to_blob(new_internal_node_index, new_internal_block.to_bytes())?;
+
+        let Some(old_parent_index) = old_leaf.parent else {
+            panic!("{key_value:?} {hash:?}")
+        };
+
+        let mut block = Block::from_bytes(
+            self.get_block_bytes(old_leaf.index)?,
+            new_internal_node_index,
+        )?;
+        block.node.parent = Some(new_internal_node_index);
+        self.insert_entry_to_blob(old_leaf.index, block.to_bytes())?;
+
+        let mut old_parent_block =
+            Block::from_bytes(self.get_block_bytes(old_parent_index)?, old_parent_index)?;
+        match old_parent_block.node.specific {
```

Comment:
this would probably be a bit simpler as a let-else.

Reply:
probably 28548288a3a4894be7d09f99f8ee584ae2508ca4